# Game of Thrones Deaths

#### Every on screen death in the Game of Thrones TV show is in this data set - details will include who, where, what and when with regards to each kill.

COLUMN NAME

name: The victim's name

allegiance: The victim's house they were associated with

season: The season of the tv show that the death occurred

episode: The episode of the tv show that the death occurred

location: The location the death occured

killer: The killer's name

killers_house: The killer's house they were associated with

method: How the killer killed their victim

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import io
# importing the openpyxsl module 
import openpyxl as xsl

In [3]:
# The excel file is loaded to wb. 
# This wb is later modified and saved as file name

wb = xsl.load_workbook('game-of-thones-deaths.xlsx')

In [4]:
wb.sheetnames

['Sheet 1 - game-of-thones-deaths']

In [12]:
df=pd.read_excel('game-of-thones-deaths.xlsx')

In [13]:
df

game-of-thones-deaths Unnamed: 1 Unnamed: 2          Unnamed: 3  \
0               Allegiance  Death No.    Episode              Killer   
1            Night's Watch          1          1        White Walker   
2            Night's Watch          2          1        White Walker   
3            Night's Watch          3          1           Ned Stark   
4                     None          4          1            Direwolf   
...                    ...        ...        ...                 ...   
2220                  None       2207          5  Daenerys Targaryen   
2221                  None       2208          5  Daenerys Targaryen   
2222                  None       2209          5  Daenerys Targaryen   
2223                  None       2210          5  Daenerys Targaryen   
2224       House Targaryen       2224          6            Jon Snow   

           Unnamed: 4       Unnamed: 5           Unnamed: 6  \
0       Killers House         Location               Method   
1                None  Beyond the Wall            Ice sword   
2                None  Beyond the Wall            Ice sword   
3         House Stark       Winterfell                Sword   
4                None       Winterfell                Teeth   
...               ...              ...                  ...   
2220  House Targaryen   King's Landing  Dragonfire (Dragon)   
2221  House Targaryen   King's Landing  Dragonfire (Dragon)   
2222  House Targaryen   King's Landing  Dragonfire (Dragon)   
2223  House Targaryen   King's Landing  Dragonfire (Dragon)   
2224      House Stark   King's Landing                Knife   

                  Unnamed: 7 Unnamed: 8  
0                       Name     Season  
1               Waymar Royce          1  
2                      Gared          1  
3                       Will          1  
4                       Stag          1  
...                      ...        ...  
2220  King's Landing Citizen          8  
2221  King's Landing Citizen          8  
2222  King's Landing Citizen          8  
2223  King's Landing Citizen          8  
2224      Daenerys Targaryen          8  

[2225 rows x 9 columns]

As 1st row here contains column names so we can convert this first row into column names with the following code.

In [14]:
new_header=df.iloc[0] #selecting first row.
df=df[1:] # treats dataset from second row onwards.
df.columns=new_header

In [15]:
df

0          Allegiance Death No. Episode              Killer    Killers House  \
1       Night's Watch         1       1        White Walker             None   
2       Night's Watch         2       1        White Walker             None   
3       Night's Watch         3       1           Ned Stark      House Stark   
4                None         4       1            Direwolf             None   
5                None         5       1                Stag             None   
...               ...       ...     ...                 ...              ...   
2220             None      2207       5  Daenerys Targaryen  House Targaryen   
2221             None      2208       5  Daenerys Targaryen  House Targaryen   
2222             None      2209       5  Daenerys Targaryen  House Targaryen   
2223             None      2210       5  Daenerys Targaryen  House Targaryen   
2224  House Targaryen      2224       6            Jon Snow      House Stark   

0            Location               Method                    Name Season  
1     Beyond the Wall            Ice sword            Waymar Royce      1  
2     Beyond the Wall            Ice sword                   Gared      1  
3          Winterfell                Sword                    Will      1  
4          Winterfell                Teeth                    Stag      1  
5          Winterfell               Antler                Direwolf      1  
...               ...                  ...                     ...    ...  
2220   King's Landing  Dragonfire (Dragon)  King's Landing Citizen      8  
2221   King's Landing  Dragonfire (Dragon)  King's Landing Citizen      8  
2222   King's Landing  Dragonfire (Dragon)  King's Landing Citizen      8  
2223   King's Landing  Dragonfire (Dragon)  King's Landing Citizen      8  
2224   King's Landing                Knife      Daenerys Targaryen      8  

[2224 rows x 9 columns]

In [47]:
df['Killer'].value_counts()

Daenerys Targaryen    1008
Cersei Lannister       199
Arya Stark              68
Wight                   54
Jon Snow                51
                      ... 
Shaggydog                1
Dickon Tarly             1
Lysa Arryn               1
Direwolf                 1
Tommen Baratheon         1
Name: Killer, Length: 129, dtype: int64

Forming a list to store counts of each Killer's kill streak.

In [55]:
list_kills=[]
list_kills.append(df['Killer'].value_counts())
list_kills[0][1]

199

For plotting a plotly.groupby we need to have number of kills the killers made as it needs to have some value to plot on.

In [25]:
list_kills_2=[]
for i in range(129):
    list_kills_2.append(list_kills[0][i])
print(list_kills_2)

[1008, 199, 68, 54, 51, 47, 41, 38, 38, 34, 32, 32, 31, 27, 27, 25, 24, 22, 20, 19, 17, 15, 14, 13, 13, 13, 11, 11, 10, 10, 10, 9, 9, 9, 8, 8, 8, 7, 7, 7, 6, 6, 5, 5, 5, 5, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


Assessing the index values of killers value counts.

https://stackoverflow.com/questions/36973387/accessing-first-column-of-pandas-value-counts

In [56]:
index_val=df['Killer'].value_counts().index

In [58]:
index_val_list=[]
for i in index_val:
    index_val_list.append(i)
print(index_val_list)

['Daenerys Targaryen', 'Cersei Lannister', 'Arya Stark', 'Wight', 'Jon Snow', 'Bolton soldier', "Night's Watch brother", 'Sons of the Harpy agent', 'Stark Soldier', 'Grey Worm', 'Dothraki man', "Baratheon of King's Landing soldier", 'Sandor "the Hound" Clegane', 'Bronn', 'Frey soldier', 'Lannister soldier', 'Baratheon of Dragonstone soldier', 'Daario Naharis', 'Tormund', 'Ramsay Bolton', 'Theon Greyjoy', 'Barristan Selmy', 'Ygritte', 'Brienne of Tarth', 'Jorah Mormont', 'Unsullied Soldier', 'Yara Greyjoy', 'Pyat Pree', 'None', 'Wildling', 'Gregor "the Mountain" Clegane', 'Jaime Lannister', 'Viserion', 'Euron Greyjoy', 'Giant Wight', 'Greyjoy Soldier', 'Styr', 'Melisandre "the Red Woman" of Asshai', 'Unsullied', 'Stannis Baratheon', 'City watch guard', 'White Walker', 'Yoren', 'Ned Stark', 'Smalljon Umber', 'Tyrion Lannister', 'Knights of the Vale', 'Thenn raider', 'Loras Tyrell', 'Tyene Sand', 'Gladiator', 'Summer', "Night's Watch mutineer", 'Arthur Dayne', 'Jory Cassel', 'Wun Wun', 'A

## Top Killers

Daenerys being the top killer has the highest count of 1008 people. No wonder how she could have ruled the world with no people in it! ;(

In [20]:
import plotly.express as px
fig=px.histogram(df, x='Killer')
fig.show()

##### using both the list to plot the figure.

In [59]:
import plotly.io as pio
data = [dict(
  type = 'scatter',
  x = index_val_list,
  y = list_kills_2,
  mode = 'markers',
  transforms = [dict(
    type = 'groupby',
    groups = index_val_list,
  )]
)]

fig_dict = dict(data=data)
pio.show(fig_dict, validate=False)

In [63]:
df['Method'].value_counts()

Dragonfire (Dragon)    986
Sword                  322
Wildfire               210
Knife                  119
Arrow                   95
                      ... 
Hands/Necklace           1
Antler                   1
Water                    1
Blunt object             1
Dragonglass              1
Name: Method, Length: 67, dtype: int64

In [69]:
list_method=[]
list_method.append(df['Method'].value_counts())
list_method[0][0]

986

In [71]:
method=[]
for i in range(67):
    method.append(list_method[0][i])
print(method)

[986, 322, 210, 119, 95, 81, 63, 48, 44, 35, 29, 23, 20, 18, 14, 9, 8, 8, 7, 7, 7, 4, 4, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]


In [72]:
index_val2=df['Method'].value_counts().index
index_val_list2=[]
for i in index_val2:
    index_val_list2.append(i)
print(index_val_list2)

['Dragonfire (Dragon)', 'Sword', 'Wildfire', 'Knife', 'Arrow', 'Spear', 'Poison', 'Axe', 'Arakh', 'Flaming arrow', 'Fire', 'Hands', 'Teeth', 'Falling', 'Unknown', 'Scythe', 'Noose', 'Burning oil', 'Club', 'Giant arrow', 'Barrel', 'Flaying', 'Hammer', 'Claws', 'Catapult', 'Boulder', 'Mace', 'Dogs', 'Old Age', 'Rubble', 'Ice sword', 'Rock', 'Cooking', 'Shield', 'Teeth (Dragon)', 'Sealed into vault', 'Claws (Dragon)', "Butcher's knife", 'Bludgeon', 'Starvation', 'Pot', 'Broken foot', 'Hanging', 'Mallet', 'Bolo', 'Rope', 'Magic', 'Mammoth foot', 'Molten gold', 'Drowning', 'Pillow', 'Tusk', 'Rat', 'Illness', 'Moon Door', 'Steak', 'Nail', 'Poison dart', 'Shadow Demon', 'Ice Javelin', 'Whip', 'Bell', 'Hands/Necklace', 'Antler', 'Water', 'Blunt object', 'Dragonglass']


In [73]:
import plotly.io as pio
data = [dict(
  type = 'scatter',
  x = index_val_list2,
  y = method,
  mode = 'markers',
  transforms = [dict(
    type = 'groupby',
    groups = index_val_list2,
  )]
)]

fig_dict = dict(data=data)
pio.show(fig_dict, validate=False)